--- a/modutils.sh	2019-04-10 07:06:59.431408000 +0300
+++ b/modutils.sh	2019-04-10 22:20:37.453354878 +0300
@@ -36,6 +36,87 @@
 		[ "$VERBOSE" != no ] && echo -n "$module "
 		eval "$LOAD_MODULE $module $args >/dev/null 2>&1"
 		loaded_modules="${loaded_modules}${module} "
+		if [ "$module" == "boxtype" ]; then
+			boxtype=`cat /proc/boxtype`
+			echo "boxtype $boxtype"
+
+			if [ $boxtype == 'bska' ] || [ $boxtype == 'bxzb' ]; then
+				echo "Init VFD white"
+				modprobe front_bska
+			else
+				echo "Init VFD black"
+				modprobe front_bsla
+			fi
+
+			mknod -m 0660 /dev/vfd c 147 0
+			echo "LOAD" >/dev/vfd
+			mknod -m 666 /dev/sci1 c 169 0
+			mknod -m 666 /dev/sci1 c 169 1
+
+			modprobe simu_button
+			mknod /dev/rc c 147 1
+			modprobe e2_proc
+			modprobe stmcore-display-stx7100
+			modprobe stmfb display0=1280x720-32@50:16m:pal:yuv:yuv
+			modprobe embxshell
+			modprobe embxmailbox mailbox0=0xb9211000:136:set2 mailbox1=0xb9212000:0
+			modprobe embxshm mailbox0=shm:0:7:0x60000000:0:16:16:0:1024:0xa4000000:0x10000000
+			modprobe mme_host transport0=shm
+
+			if [ $boxtype == 'bsla' ] || [ $boxtype == 'bzzb' ]; then
+				modprobe cooler
+			fi
+
+			if [ $boxtype == 'bxzb' ]; then
+				modprobe avs type=fake_avs
+			else
+				modprobe avs type=stv6412
+			fi
+
+			modprobe mmelog
+			modprobe ksound
+			modprobe p2div64
+			modprobe stm_v4l2
+			modprobe stmvout
+			modprobe stmvbi
+			modprobe pseudocard
+
+			if [ $boxtype == 'bska' ] || [ $boxtype == 'bxzb' ]; then
+				modprobe fe-core
+			elif [ $boxtype == 'bsla' ]; then
+				modprobe fe-core-twin1
+			else
+				modprobe fe-core-twin2
+			fi
+
+			modprobe pti waitMS=20 videoMem=4096
+			modprobe stm_monitor
+			modprobe stmsysfs
+
+			if [ $boxtype == 'bzzb' ]; then
+				modprobe stmdvb TSIS_mode=serial NIMS=twin
+			elif [ $boxtype == 'bsla' ]; then
+				modprobe stmdvb TSIS_mode=parallel NIMS=twin
+			else
+				modprobe stmdvb TSIS_mode=parallel NIMS=single
+			fi
+
+			modprobe player2 discardlateframe=0
+			modprobe sth264pp
+			modprobe stmalloc
+			modprobe platform
+			modprobe silencegen
+			#smartcard parametry: type=bxzb, type=bska, type=bsla, mode_sci1=nbox, mode_sci1=tda8004
+			if [ $boxtype == 'bxzb' ]; then
+				modprobe smartcard type=bxzb mode_sci1=nbox
+			else
+				modprobe smartcard type=bska mode_sci1=nbox
+			fi
+
+			modprobe bpamem
+			modprobe cec_ctrl
+		fi
+		[ "$module" == "zram" ] && echo "16777216" > /sys/block/zram0/disksize # Use 16Mb for swap in ram
 	done
 }
 
@@ -49,4 +130,13 @@
 done
 [ "$VERBOSE" != no ] && echo
 
+echo "1" > /proc/sys/kernel/printk #for more log change level or remake line
+
+# Enable swap in ram
+if [ -e /dev/zram0 ]; then
+	[ "$VERBOSE" != no ] && echo "Enable swap in ram"
+	mkswap /dev/zram0
+	swapon -p 100 /dev/zram0
+fi
+
 exit 0
