diff -Naur apps/tools/evremote2/global.c apps-PersianPrince/tools/evremote2/global.c
--- apps/tools/evremote2/global.c	2019-01-12 13:54:32.162690614 +0330
+++ apps-PersianPrince/tools/evremote2/global.c	2019-01-12 13:53:24.900359867 +0330
@@ -37,7 +37,7 @@
 #include "global.h"
 #include "map.h"
 
-#define DEVICENAME "TDT RC event driver"
+#define DEVICENAME "SH4 RC event driver"
 char eventPath[] = "/dev/input/event0";
 
 // Checks which event device is created by simubutton.ko
diff -Naur apps/tools/wait4button/global.c apps-PersianPrince/tools/wait4button/global.c
--- apps/tools/wait4button/global.c	2019-01-12 13:54:32.234688829 +0330
+++ apps-PersianPrince/tools/wait4button/global.c	2019-01-12 13:52:12.578158476 +0330
@@ -37,7 +37,7 @@
 #include "global.h"
 #include "map.h"
 
-#define DEVICENAME "TDT RC event driver"
+#define DEVICENAME "SH4 RC event driver"
 char eventPath[] = "/dev/input/event0";
 
 // Checks which event device is created by simubutton.ko
