diff -Naur git/tools/evremote2/global.c git-PersianPrince/tools/evremote2/global.c
--- git/tools/evremote2/global.c	2019-01-12 14:23:40.326869000 +0330
+++ git-PersianPrince/tools/evremote2/global.c	2019-01-12 14:26:12.955795447 +0330
@@ -37,7 +37,7 @@
 #include "global.h"
 #include "map.h"
 
-#define DEVICENAME "TDT RC event driver"
+#define DEVICENAME "SH4 RC event driver"
 char eventPath[] = "/dev/input/event0";
 
 // Checks which event device is created by simubutton.ko
